publisher.go

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2019 Dgraph Labs, Inc. and Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package badger
+
+import (
+	"bytes"
+	"sync"
+
+	"github.com/dgraph-io/badger/pb"
+	"github.com/dgraph-io/badger/y"
+)
+
+type subscriber struct {
+	prefix []byte
+	sendCh chan<- *pb.KVList
+}
+
+type publisher struct {
+	sync.Mutex
+	pubCh       chan []*request
+	subscribers map[uint64]subscriber
+	lastID      uint64
+}
+
+func newPublisher() *publisher {
+	return &publisher{
+		pubCh:       make(chan []*request, 10000),
+		subscribers: make(map[uint64]subscriber),
+		lastID:      0,
+	}
+}
+
+func (p *publisher) listenForUpdates(c *y.Closer) {
+	defer func() {
+		c.Done()
+		p.cleanSubscribers()
+	}()
+
+listen:
+	for {
+		reqs := []*request{}
+		select {
+		case <-c.HasBeenClosed():
+			break listen
+		case r := <-p.pubCh:
+			reqs = append(reqs, r...)
+		}
+	drainer:
```

Comment:
I'm not sure whether this is really a good idea. It delays sending the updates, possibly blocking forever if rate of publish > rate of drain. We should potentially cap the size of updates we send.

Reply:
writes are getting batched with a cap at `doWrites`. I think no need to drain since it is already batched. I could just push after receiving. what do you think? should I set a cap or just push.